onnxruntime/core/providers/cpu/tensor/cast_op.cc

```diff
@@ -1,193 +1,231 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
+#include <cstddef>
 #include <iomanip>
 #include <sstream>
+
+#include "boost/mp11.hpp"
+
+#include "gsl/gsl"
+
 #include "core/common/common.h"
+#include "core/common/type_list.h"
+#include "core/framework/data_types.h"
+#include "core/framework/data_types_internal.h"
 #include "core/framework/op_kernel.h"
 #include "core/providers/cpu/tensor/utils.h"
-#include "core/util/math.h"
+#include "core/providers/op_kernel_type_control.h"
 #include "core/util/math_cpuonly.h"
+
 #include "Eigen/src/Core/arch/Default/Half.h"
 
 #if defined(_M_AMD64)
 #include "core/mlas/inc/mlas.h"
 #endif
 
-// FUTURE:
-// Float16 and String have expensive special cased handling. Enable by default, but provide an easy way to disable
-// in the future if needed. Disabling both saves ~50KB.
-#define CAST_FLOAT16_ENABLED
-#define CAST_STRING_ENABLED
-
 using namespace ONNX_NAMESPACE;
+using namespace boost::mp11;
+
 namespace onnxruntime {
 
+namespace op_kernel_type_control {
+ORT_SPECIFY_OP_KERNEL_ARG_SUPPORTED_TYPES(
+    Cast, Input, 0,
+    bool,
+    float, double,
+    uint8_t, uint16_t, uint32_t, uint64_t,
+    int8_t, int16_t, int32_t, int64_t,
+    MLFloat16, BFloat16,
+    std::string);
+
+ORT_SPECIFY_OP_KERNEL_ARG_SUPPORTED_TYPES(
+    Cast, Output, 0,
+    bool,
+    float, double,
+    uint8_t, uint16_t, uint32_t, uint64_t,
+    int8_t, int16_t, int32_t, int64_t,
+    MLFloat16, BFloat16,
+    std::string);
+}  // namespace op_kernel_type_control
+
 namespace {
-template <typename SrcType, typename DstType>
-inline void CastData(const Tensor& in, Tensor& out, const TensorShape& shape) {
-  ptrdiff_t shape_size = gsl::narrow<ptrdiff_t>(shape.Size());
-  auto in_vector = ConstEigenVectorMap<SrcType>(in.Data<SrcType>(), shape_size);
-  auto output_vector = EigenVectorMap<DstType>(out.MutableData<DstType>(), shape_size);
-  output_vector = in_vector.template cast<DstType>();
```

Comment:
Any potential perf impact from not using Eigen for this? Not sure if Eigen has any optimizations for this operation that we'd be losing.

Reply:
will try to do some measurements

---
In reply to: [564887790](https://github.com/microsoft/onnxruntime/pull/6466#discussion_r564887790) [](ancestors = 564887790)

Reply:
ongoing, will do any necessary fixes in a future PR

---
In reply to: [564929687](https://github.com/microsoft/onnxruntime/pull/6466#discussion_r564929687) [](ancestors = 564929687,564887790)